llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -836,7 +836,7 @@ class InstructionsState {
     return getOpcode() == CheckedOpcode || getAltOpcode() == CheckedOpcode;
   }
 
-  InstructionsState() = delete;
+  InstructionsState() = default;
```

Comment:
This is the change that should be avoided

Reply:
The problem is the constructor of `TreeEntry`.
```
  struct TreeEntry {
    using VecTreeTy = SmallVector<std::unique_ptr<TreeEntry>, 8>;
    TreeEntry(VecTreeTy &Container) : Container(Container) {}
```
`Container` is just used to dump the SLP tree. See 95da05c3f5064c733dc486a43f262a55deb03dc5.

Reply:
Why do you want to store state at all?

Reply:
1. From `setOperations` we can know it should store InstructionsState.
```
    void setOperations(const InstructionsState &S) {
      MainOp = S.getMainOp();
      AltOp = S.getAltOp();
    }
```
2. The `TreeEntry` reimplemented all of the functions of `InstructionsState`. If we need to add some functions in `InstructionsState`, we may need to reimplement it in `TreeEntry` again.
3. I expect to add some functions in `InstructionsState` to solve the performance issue for https://github.com/llvm/llvm-project/pull/112181.

Reply:
> 1. From `setOperations` we can know it should store InstructionsState.
> 
> ```
>     void setOperations(const InstructionsState &S) {
>       MainOp = S.getMainOp();
>       AltOp = S.getAltOp();
>     }
> ```
> 
> 2. The `TreeEntry` reimplemented all of the functions of `InstructionsState`. If we need to add some functions in `InstructionsState`, we may need to reimplement it in `TreeEntry` again.
> 3. I expect to add some functions in `InstructionsState` to solve the performance issue for [[SLP] Make getSameOpcode support different instructions if they have same semantics. #112181](https://github.com/llvm/llvm-project/pull/112181).

And how does current infrastructure prevent it?

Reply:
The problem is when we call `newTreeEntry`, we don't provide the `MainOp` and `AltOp`.
```
  TreeEntry *newTreeEntry(ArrayRef<Value *> VL,
                          TreeEntry::EntryState EntryState,
                          std::optional<ScheduleData *> Bundle,
                          const InstructionsState &S,
                          const EdgeInfo &UserTreeIdx,
                          ArrayRef<int> ReuseShuffleIndices = {},
                          ArrayRef<unsigned> ReorderIndices = {}) {
    assert(((!Bundle && EntryState == TreeEntry::NeedToGather) ||
            (Bundle && EntryState != TreeEntry::NeedToGather)) &&
           "Need to vectorize gather entry?");
    // Gathered loads still gathered? Do not create entry, use the original one.
    if (GatheredLoadsEntriesFirst.has_value() &&
        EntryState == TreeEntry::NeedToGather &&
        S.getOpcode() == Instruction::Load && UserTreeIdx.EdgeIdx == UINT_MAX &&
        !UserTreeIdx.UserTE)
      return nullptr;
    VectorizableTree.push_back(std::make_unique<TreeEntry>(VectorizableTree));
    TreeEntry *Last = VectorizableTree.back().get();
```
It will make `TreeEntry` call the default constructor of `InstructionsState`. However, it is deleted.

This patch will make `InstructionsState` can be constructed by default constructor.

Reply:
It is provided in `const InstructionsState &S,`